src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotFirecloudGroupStep.java

```diff
@@ -0,0 +1,39 @@
+package bio.terra.service.snapshot.flight.create;
+
+import bio.terra.model.FirecloudGroupModel;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.snapshot.flight.SnapshotWorkingMapKeys;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import java.util.UUID;
+
+public class CreateSnapshotFirecloudGroupStep implements Step {
+  private final IamService iamService;
+  private final UUID snapshotId;
+
+  public CreateSnapshotFirecloudGroupStep(IamService iamService, UUID snapshotId) {
+    this.iamService = iamService;
+    this.snapshotId = snapshotId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FirecloudGroupModel created = iamService.createFirecloudGroup(snapshotId.toString());
+    context
+        .getWorkingMap()
+        .put(SnapshotWorkingMapKeys.SNAPSHOT_FIRECLOUD_GROUP, created.getGroupName());
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    String createdGroup =
+        context.getWorkingMap().get(SnapshotWorkingMapKeys.SNAPSHOT_FIRECLOUD_GROUP, String.class);
+    if (createdGroup != null) {
+      iamService.deleteGroup(createdGroup);
```

Comment:
I forget the exact reasoning, but I think we try to not write a property in `do` and then read it in the same step's `undo`. I think the hazard in this case is that the group gets created but stairway fails to write the map to the database, and then the operation can't be undone. Splitting it up into two steps, where the first step creates the unique name, is usually how it's done.

Reply:
I was reusing the logic used to create the firecloud group for datasets with duos ids, but I'd be happy to talk through this approach with you.

Reply:
I think this is okay as is - We're not going to know what the group name is until the call succeeds (Since we have two options for the name).  The createdGroup null check handles the case that the firecloud group wasn't written to the map so the flight shouldn't dismally fail. 
Yes, it's true that the group create could succeed and then we don't successful write the name to the map. This seems like an unlikely scenario and I think it is not a huge deal if we leave one or two stray groups over the course of the life of this code. If we're somehow leaving a bunch of stray groups, then I'd be concerned.